### PR TITLE
in details panel, count connections only once

### LINKED ISSUE #1842: details panel connection counts are too high
To reproduce
```
$ scope launch
1c41ae18a677d54463e23e3adc5e63516b7409688634db79a9d8df5dcc9d635e
Weave Scope is reachable at the following URL(s):
  [...]
$ google-chrome http://localhost:4040
$ docker run --rm --name=foo -ti alpine /bin/sh
/ # IP=$(getent hosts cloud.weave.works | cut -d ' ' -f 1)
/ # echo $IP
52.3.116.14
/ # nc $IP 80
```
Wait for the 'foo' container and connection to the Internet node to show up in scope and then bring up the details panel on the 'foo' node. You should see something like
screenshot_2016-09-01_22-43-19

Notice that it says there are *two* connections from the node to the Internet, when in reality there is only one.

This doesn't happen all the time, unfortunately, though right now for me it happens more often than not. Sometimes I initially see a count of 1 and it then increases to 2. And then sometimes it drops back to 1.

### PR DESCRIPTION
**Pull request by rade**
Fixes #1842

### CODE REVIEW

File: render/detailed/connections.go [hunk 1]
```diff
@@ -85,25 +85,111 @@ func (row connection) ID() string {
...
+	// ensure we only count them once.
+	//
+	// There is some arbitrariness here: We may see the same
+	// connection multiple times, with different destination
+	// endpoints, due to DNATing; the (destination) port under which
+	// we track that connection is determined by the first destination
+	// endpoint we encounter.
+	//
+	// The arbitrariness is mitigated by the calling code by
+	// subsituting the destination endpoint with the "copy_of" if
```
Comment (2opremio) [inline render/detailed/connections.go]: typo: substituting

File: render/detailed/connections.go [hunk 2]
```diff
@@ -85,25 +85,111 @@ func (row connection) ID() string {
 	return fmt.Sprintf("%s:%s-%s:%s-%s", row.remoteNodeID, row.remoteAddr, row.localNodeID, row.localAddr, row.port)
 }
 
+type connections struct {
```
Comment (2opremio) [inline render/detailed/connections.go]: `connection`, `connections` and `[]Connection` make the code a a bit hard to read. I would consider renaming this type to something more explicit like `connectionCounts`

File: render/detailed/connections.go [hunk 3]
```diff
@@ -117,36 +203,37 @@ func incomingConnectionsSummary(topologyID string, r report.Report, n report.Nod
...
 
 		for _, localEndpoint := range localEndpoints {
 			_, localAddr, _, ok := report.ParseEndpointNodeID(localEndpoint.ID)
 			if !ok {
 				continue
 			}
-
 			for _, remoteEndpointID := range localEndpoint.Adjacency.Intersection(remoteEndpointIDs) {
+				if remoteEndpointIDCopy, ok := remoteEndpointIDCopies[remoteEndpointID]; ok {
+					remoteEndpointID = remoteEndpointIDCopy
```
Comment (2opremio) [inline render/detailed/connections.go]: I think a comment is in order here to indicate we need this to make the destination unique.

File: render/detailed/connections.go (hunk 2 above)
Comment (rade) [inline render/detailed/connections.go]: That leads to stuttering. `connectionCounts.counts`.

Comment (2opremio) [inline render/detailed/connections.go]: I think that the tradeoff is worth it though.

File: render/detailed/connections.go (hunk 1 above)
Comment (rade) [inline render/detailed/connections.go]: fixed

File: render/detailed/connections.go (hunk 2 above)
Comment (rade) [inline render/detailed/connections.go]: ok. fixed.

File: render/detailed/connections.go (hunk 3 above)
Comment (rade) [inline render/detailed/connections.go]: fixed, by creating a helper function with a comment.

**Comment by rade** [comment]
@2opremio PTAL

File: render/detailed/connections.go [hunk 4]
```diff
@@ -174,65 +246,36 @@ func endpointChildrenOf(n report.Node) []report.Node {
...
+// canonicalEndpointID returns the original endpoint ID of which id is
+// a "copy_of" (due to NATing), or, if the id is not a copy, the id
+// itself.
+//
+// This is used for determining a unique destination endpoint ID for a
+// connection, removing any arbitrariness in the destination port we
+// are associating with the connection when it is encountered multiple
+// times in the topology (with different destination endpoints, due to
+// DNATing).
+func canonicalEndpointID(copies map[string]string, id string) string {
```
Comment (2opremio) [inline render/detailed/connections.go]: So much nicer, thanks!

**Comment by 2opremio** [comment]
LGTM

File: render/detailed/connections.go [hunk 5]
```diff
@@ -85,25 +85,98 @@ func (row connection) ID() string {
...
+	connectionID := sourceEndpoint.ID
+	if copySourceEndpointID, _, ok := sourceEndpoint.Latest.LookupEntry("copy_of"); ok {
+		connectionID = copySourceEndpointID
+	}
+	if _, ok := c.counted[connectionID]; ok {
+		return
+	}
+	c.counted[connectionID] = struct{}{}
+	key := newConnection(n, node, port, endpointID, localAddr)
+	c.counts[key] = c.counts[key] + 1
```
Comment (peterbourgon) [inline render/detailed/connections.go]:
```go
c.counts[key]++
```

Comment (rade) [inline render/detailed/connections.go]: I didn't write that code.

Comment (rade) [inline render/detailed/connections.go]: fixed in 94cd73c72

**Pull request merged**